apps/studio/src/server/modules/permissions/permissions.service.ts

```diff
@@ -0,0 +1,63 @@
+import { AbilityBuilder, createMongoAbility, PureAbility } from "@casl/ability"
+
+import { db, Resource as RawResource, RoleType } from "../database"
+
+type Resource = Pick<RawResource, "parentId">
+type AllowedResourceActions = (typeof ALL_ACTIONS)[number]
+type Subjects = "Resource" | Resource
+
+export const CRUD_ACTIONS = ["create", "read", "update", "delete"] as const
+export const ALL_ACTIONS = [...CRUD_ACTIONS, "move"] as const
+
+export type ResourceAbility = PureAbility<[AllowedResourceActions, Subjects]>
+
+interface PermissionsProps {
+  userId: string
+  siteId: number
+  resourceId: string | null
+}
+export const definePermissionsFor = async ({
+  userId,
+  siteId,
+  resourceId,
+}: PermissionsProps) => {
```

Comment:
from review w kr: we still need a way for the caller or this method to know if it is site-wide or root level permissions. from chin: i was thinking of defining a method `getRootPerms` that retrieves site wide perms or, if not possible, root level

Reply:
another way we could do this is also via growth book flagging. so actually next time we always need to union the list of permissions upwards via CTE :sadge:

Reply:
asked about using growthbook as a feature flag - karrui was against it. this is because we don't want business logic to be tied up in growthbook, i think his point was that root level permissions are an unimplemented subset of our current feature and we should not restrict access to admins

Reply:
>  if it is site-wide or root level permissions
This is why I suggested we need to store the context for a site if it is using site wide or root level perms, likely in Site. But this will add 1 more query for us to check this entry

Agreed we shouldnt tie up biz logic within GB too